bootloader/piboot.go

```diff
@@ -0,0 +1,456 @@
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
+	ubuntuBootDir = runMntDir + "ubuntu-boot/"
+)
+
+type piboot struct {
+	rootdir          string
+	basedir          string
+	prepareImageTime bool
+}
+
+func (p *piboot) setDefaults() {
+	p.basedir = "/boot/piboot/"
+}
+
+func (p *piboot) processBlOpts(blOpts *Options) {
+	if blOpts != nil {
+		p.prepareImageTime = blOpts.PrepareImageTime
+		switch {
+		case blOpts.Role == RoleRecovery || blOpts.NoSlashBoot:
+			// RoleRecovery or NoSlashBoot imply we use
+			// the environment file in /piboot/ubuntu as
+			// it exists on the partition directly
+			p.basedir = pibootPartFolder
+		}
+	}
+}
+
+// newPiboot creates a new Piboot bootloader object
+func newPiboot(rootdir string, blOpts *Options) Bootloader {
+	p := &piboot{
+		rootdir: rootdir,
+	}
+	p.setDefaults()
+	p.processBlOpts(blOpts)
+	return p
+}
+
+func (p *piboot) Name() string {
+	return "piboot"
+}
+
+func (p *piboot) dir() string {
+	if p.rootdir == "" {
+		panic("internal error: unset rootdir")
+	}
+	return filepath.Join(p.rootdir, p.basedir)
+}
+
+func (p *piboot) envFile() string {
+	return filepath.Join(p.dir(), pibootCfgFilename)
+}
+
+// piboot enabled if env file exists
+func (p *piboot) Present() (bool, error) {
+	return osutil.FileExists(p.envFile()), nil
+}
+
+// Variables stored in ubuntu-seed:
+//   snapd_recovery_system
+//   snapd_recovery_mode
+//   snapd_recovery_kernel
+// Variables stored in ubuntu-boot:
+//   kernel_status
+//   snap_kernel
+//   snap_try_kernel
+//   snapd_extra_cmdline_args
+//   snapd_full_cmdline_args
+//   recovery_system_status
+//   try_recovery_system
+func (p *piboot) SetBootVars(values map[string]string) error {
+	env := pibootenv.NewEnv(p.envFile())
+	if err := env.Load(); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	dirty := false
+	modeChanged := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirty = true
+		if k == "snapd_recovery_mode" {
+			modeChanged = true
+		}
+	}
+
+	if dirty {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	// Configure the bootloader when recovery mode is changed
+	// (kernel command line changes)
+	if modeChanged {
+		if err := p.loadAndApplyConfig(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) loadAndApplyConfig() error {
+	var prefix, cfgDir, dstDir string
+	env := pibootenv.NewEnv(p.envFile())
+	if p.prepareImageTime {
+		if err := env.Load(); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		prefix = filepath.Join("/systems", env.Get("snapd_recovery_system"),
+			"kernel")
+		cfgDir = p.rootdir
+		dstDir = filepath.Join(p.rootdir, prefix)
+	} else {
+		// Make sure to load all config files
+		mntDirs := []string{ubuntuSeedDir, ubuntuBootDir}
+		for _, dir := range mntDirs {
+			cfgPath := filepath.Join(dir, pibootPartFolder, pibootCfgFilename)
```

Comment:
for readability, maybe it would be better to consider pibootCfgPath instead of cfgPath